src/librustc_hir/hir.rs

```diff
@@ -638,6 +638,8 @@ pub struct Crate<'hir> {
     /// A list of proc macro HirIds, written out in the order in which
     /// they are declared in the static array generated by proc_macro_harness.
     pub proc_macros: Vec<HirId>,
+
+    pub trait_map: BTreeMap<HirId, Vec<TraitCandidate<HirId>>>,
```

Comment:
To clarify, the logic is that `Resolver::into_outputs` cannot lower `NodeId`s to `HirId`s if we remove the relevant conversion maps from `Definitions`, so we have to move it to creation of this map to some place where these conversion maps are still available, i.e. AST lowering.

Reply:
`trait_map` can use `TraitCandidate<LocalDefId>` both in resolver and later.
Or rather `TraitCandidate` can be done non-generic and use `LocalDefId` directly.

Reply:
Could you remove the `type TraitMap` alias as well?

Reply:
So `trait_map` cannot use `LocalDefId`, it really is storing `HirId` that don't have associated `LocalDefId`, at least either for the key or the value (can't remember which, maybe both).

Reply:
@petrochenkov if you have any advises on how to use `LocalDefId` in the `trait_map` i am all hear because i am a bit stuck here.

Reply:
https://github.com/rust-lang/rust/pull/73357
I'm sorry.

Reply:
I tried to check whether are any problems with converting these in practice and there was none, so I just sent a PR.

Reply:
I am not sure I am following. In the PR you just sent, you are still calling `node_id_to_hir_id` to convert the keys whereas that function would be removed alongside the backing conversion table with this PR.

Reply:
>`trait_map` can use `TraitCandidate<LocalDefId>` both in resolver and later.

^ The original suggestion.
It's about values of the trait map, not keys.

Keys still have to be fine-grained and use `NodeId`s early and `HirId`s later, so for keys this PR does the right thing.

Reply:
Ok sorry for the misunderstanding. I am happy to land your PR first and rebase this one on it.